### ISSUE TITLE
Matrix multiplication error

### ISSUE THREAD

**Issue by andresgit** [tinygrad/tinygrad#3271]
I'm on the latest master running this code:

```python
from tinygrad import Tensor
m1 = Tensor.rand((4096,599*64), device="cuda").realize()
m2 = Tensor.rand((599*64,1024), device="cuda").realize()
res = m1@m2
res.realize()
```
It gives me this error:
```python
Traceback (most recent call last):
  File "/workspace/temp.py", line 5, in <module>
    res.realize()
  File "/workspace/tinygrad/tinygrad/tensor.py", line 116, in realize
    run_schedule(self.lazydata.schedule())
  File "/workspace/tinygrad/tinygrad/realize.py", line 48, in run_schedule
    prg = lower_schedule_item(si)
  File "/workspace/tinygrad/tinygrad/realize.py", line 39, in lower_schedule_item
    return Device[si.out.device].get_runner(si.ast)
  File "/workspace/tinygrad/tinygrad/device.py", line 349, in get_runner
    def get_runner(self, ast:LazyOp) -> CompiledASTRunner: return self.to_program(self.get_linearizer(ast))
  File "/workspace/tinygrad/tinygrad/device.py", line 322, in to_program
    return CompiledASTRunner(k.ast, k.name, self.compiler.render(to_function_name(k.name), k.uops), self, k.global_size, k.local_size)
  File "/workspace/tinygrad/tinygrad/device.py", line 286, in __init__
    self.lib, self.clprg = lib, self.device.runtime(self.name, lib)
  File "/workspace/tinygrad/tinygrad/runtime/ops_cuda.py", line 57, in __init__
    self.module = init_c_var(cuda.CUmodule(), lambda x: check(cuda.cuModuleLoadData(ctypes.byref(x), lib)))
  File "/workspace/tinygrad/tinygrad/helpers.py", line 209, in init_c_var
    def init_c_var(ctypes_var, creat_cb): return (creat_cb(ctypes_var), ctypes_var)[1]
  File "/workspace/tinygrad/tinygrad/runtime/ops_cuda.py", line 57, in <lambda>
    self.module = init_c_var(cuda.CUmodule(), lambda x: check(cuda.cuModuleLoadData(ctypes.byref(x), lib)))
  File "/workspace/tinygrad/tinygrad/runtime/ops_cuda.py", line 28, in check
    if status != 0: raise RuntimeError(f"CUDA Error {status}, {ctypes.string_at(init_c_var(ctypes.POINTER(ctypes.c_char)(), lambda x: cuda.cuGetErrorString(status, ctypes.byref(x)))).decode()}")  # noqa: E501
RuntimeError: CUDA Error 700, an illegal memory access was encountered
Exception ignored in: <function CUDAProgram.__del__ at 0x7f179a5fa830>
Traceback (most recent call last):
  File "/workspace/tinygrad/tinygrad/runtime/ops_cuda.py", line 62, in __del__
  File "/workspace/tinygrad/tinygrad/runtime/ops_cuda.py", line 28, in check
RuntimeError: CUDA Error 700, an illegal memory access was encountered
```
If I for example change the 599\*64 to 599\*32 (for both m1 and m2) then it works.

**Comment by Cheapshot003** [comment]
maybe you're running out of memory?

**Comment by shawnha** [comment]
Can replicate this on my machine as well. Have a 2060 6GB. *64 should only be using ~765 MB no? And then *32 works using only ~390 MB.

**Comment by andresgit** [comment]
The issue seems to be the first kernel run (but it just shows the error from the next command that checks cuda return status). The first kernel is:

```
  0 ━┳ STORE MemBuffer(idx=0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(4096, 1024, 599, 1), strides=(613376, 599, 1, 0), offset=0, mask=None, contiguous=True),)))
  1  ┗━┳ SUM (4096, 1024, 599, 1)
  2    ┗━┳ MUL 
  3      ┣━━ LOAD MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(4096, 1024, 599, 64), strides=(38336, 0, 64, 1), offset=0, mask=None, contiguous=False),)))
  4      ┗━━ LOAD MemBuffer(idx=2, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(4096, 1024, 599, 64), strides=(0, 1, 65536, 1024), offset=0, mask=None, contiguous=False),)))
```

```cuda
extern "C" __global__ void r_128_16_599_8_16_16_4_4_4(float* data0, const float* data1, const float* data2) {
  int gidx0 = blockIdx.z; /* 128 */
  int gidx1 = blockIdx.y; /* 16 */
  int gidx2 = blockIdx.x; /* 599 */
  int lidx3 = threadIdx.y; /* 8 */
  int lidx4 = threadIdx.x; /* 16 */
  float acc0 = 0.0f;
  float acc1 = 0.0f;
  float acc2 = 0.0f;
  float acc3 = 0.0f;
  float acc4 = 0.0f;
  float acc5 = 0.0f;
  float acc6 = 0.0f;
  float acc7 = 0.0f;
  float acc8 = 0.0f;
  float acc9 = 0.0f;
  float acc10 = 0.0f;
  float acc11 = 0.0f;
  float acc12 = 0.0f;
  float acc13 = 0.0f;
  float acc14 = 0.0f;
  float acc15 = 0.0f;
  int alu0 = ((gidx0*19628032)+(gidx1*38336)+gidx2+(lidx3*2453504)+(lidx4*2396));
  for (int ridx0 = 0; ridx0 < 16; ridx0++) {
    int alu1 = ((gidx0*1226752)+(gidx2*64)+(lidx3*153344)+(ridx0*4));
    float4 val0 = (float4)(*((float4*)(data1+alu1)));
    float4 val1 = (float4)(*((float4*)(data1+alu1+38336)));
    float4 val2 = (float4)(*((float4*)(data1+alu1+76672)));
    float4 val3 = (float4)(*((float4*)(data1+alu1+115008)));
    int alu2 = ((gidx1*64)+(gidx2*65536)+(lidx4*4)+(ridx0*4096));
    float4 val4 = (float4)(*((float4*)(data2+alu2)));
    float4 val5 = (float4)(*((float4*)(data2+alu2+1024)));
    float4 val6 = (float4)(*((float4*)(data2+alu2+2048)));
    float4 val7 = (float4)(*((float4*)(data2+alu2+3072)));
    acc0 = (((val0).w*(val7).x)+(((val0).z*(val6).x)+(((val0).y*(val5).x)+(((val0).x*(val4).x)+acc0))));
    acc1 = (((val0).w*(val7).y)+(((val0).z*(val6).y)+(((val0).y*(val5).y)+(((val0).x*(val4).y)+acc1))));
    acc2 = (((val0).w*(val7).z)+(((val0).z*(val6).z)+(((val0).y*(val5).z)+(((val0).x*(val4).z)+acc2))));
    acc3 = (((val0).w*(val7).w)+(((val0).z*(val6).w)+(((val0).y*(val5).w)+(((val0).x*(val4).w)+acc3))));
    acc4 = (((val1).w*(val7).x)+(((val1).z*(val6).x)+(((val1).y*(val5).x)+(((val1).x*(val4).x)+acc4))));
    acc5 = (((val1).w*(val7).y)+(((val1).z*(val6).y)+(((val1).y*(val5).y)+(((val1).x*(val4).y)+acc5))));
    acc6 = (((val1).w*(val7).z)+(((val1).z*(val6).z)+(((val1).y*(val5).z)+(((val1).x*(val4).z)+acc6))));
    acc7 = (((val1).w*(val7).w)+(((val1).z*(val6).w)+(((val1).y*(val5).w)+(((val1).x*(val4).w)+acc7))));
    acc8 = (((val2).w*(val7).x)+(((val2).z*(val6).x)+(((val2).y*(val5).x)+(((val2).x*(val4).x)+acc8))));
    acc9 = (((val2).w*(val7).y)+(((val2).z*(val6).y)+(((val2).y*(val5).y)+(((val2).x*(val4).y)+acc9))));
    acc10 = (((val2).w*(val7).z)+(((val2).z*(val6).z)+(((val2).y*(val5).z)+(((val2).x*(val4).z)+acc10))));
    acc11 = (((val2).w*(val7).w)+(((val2).z*(val6).w)+(((val2).y*(val5).w)+(((val2).x*(val4).w)+acc11))));
    acc12 = (((val3).w*(val7).x)+(((val3).z*(val6).x)+(((val3).y*(val5).x)+(((val3).x*(val4).x)+acc12))));
    acc13 = (((val3).w*(val7).y)+(((val3).z*(val6).y)+(((val3).y*(val5).y)+(((val3).x*(val4).y)+acc13))));
    acc14 = (((val3).w*(val7).z)+(((val3).z*(val6).z)+(((val3).y*(val5).z)+(((val3).x*(val4).z)+acc14))));
    acc15 = (((val3).w*(val7).w)+(((val3).z*(val6).w)+(((val3).y*(val5).w)+(((val3).x*(val4).w)+acc15))));
  }
  data0[alu0] = acc0;
  data0[alu0+599] = acc1;
  data0[alu0+1198] = acc2;
  data0[alu0+1797] = acc3;
  data0[alu0+613376] = acc4;
  data0[alu0+613975] = acc5;
  data0[alu0+614574] = acc6;
  data0[alu0+615173] = acc7;
  data0[alu0+1226752] = acc8;
  data0[alu0+1227351] = acc9;
  data0[alu0+1227950] = acc10;
  data0[alu0+1228549] = acc11;
  data0[alu0+1840128] = acc12;
  data0[alu0+1840727] = acc13;
  data0[alu0+1841326] = acc14;
  data0[alu0+1841925] = acc15;
}
```

The output array has 4096\*1024\*599=2^31.226 floats and the `int alu0` overflows, becomes negative and seems to cause the memory error.

**Comment by biniyam69** [comment]
I think it might be because you're out of memory. What is the size of your gpu?

**Comment by chenyuxyz** [comment]
closing this. for now you gets an error if index range is greater than int32 max, and we track the optional int64 fix in #3961
